xdsl/transforms/convert_memref_stream_to_snitch_stream.py

```diff
@@ -70,6 +85,133 @@ def match_and_rewrite(
         )
 
 
+class StreamOpLowering(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: memref_stream.StreamingRegionOp, rewriter: PatternRewriter
+    ) -> None:
+        operand_types = tuple(
+            cast(memref.MemRefType[Attribute], value_type)
+            for value in op.operands
+            if isinstance(value_type := value.type, memref.MemRefType)
+        )
+        el_types = tuple(operand_type.element_type for operand_type in operand_types)
+        if not all(el_type == builtin.f64 for el_type in el_types):
+            # Only support f64 streams for now
+            return
+        bytes_per_element = 8
+        shapes = tuple(operand_type.get_shape() for operand_type in operand_types)
+        all_strides = tuple(
+            strides_for_affine_map(map.data, shape, bytes_per_element)
+            for map, shape in zip(op.indexing_maps, shapes, strict=True)
+        )
+        bounds = tuple(bound.data for bound in op.bounds)
+        stride_patterns = tuple(
+            snitch_stream.StridePattern.from_bounds_and_strides(bounds, strides)
+            for strides in all_strides
+        )
+        new_operands = cast_operands_to_regs(rewriter)
+        new_inputs = new_operands[: len(op.inputs)]
+        new_outputs = new_operands[len(op.inputs) :]
+        freg = riscv.FloatRegisterType.unallocated()
+
+        rewriter.replace_matched_op(
+            new_op := snitch_stream.StreamingRegionOp(
+                new_inputs,
+                new_outputs,
+                ArrayAttr(stride_patterns),
+                rewriter.move_region_contents_to_new_regions(op.body),
+            )
+        )
+
+        new_body = new_op.body.block
+
+        input_stream_types = (stream.ReadableStreamType(freg),) * len(op.inputs)
+        output_stream_types = (stream.WritableStreamType(freg),) * len(op.outputs)
+        stream_types = input_stream_types + output_stream_types
+        for i in reversed(range(len(stream_types))):
+            arg = new_body.args[i]
+            stream_type = stream_types[i]
+            rewriter.insert_op_at_start(
+                cast_op := builtin.UnrealizedConversionCastOp.get((arg,), (arg.type,)),
+                new_body,
+            )
+            arg.replace_by(cast_op.results[0])
+            cast_op.operands = (arg,)
+            rewriter.modify_block_argument_type(arg, stream_type)
+
+
+def offset_map_from_shape(shape: Sequence[int], factor: int) -> AffineMap:
+    """
+    Given a list of lengths for each dimension of a memref, and the number of bytes per
+    element, returns the map from indices to an offset in bytes in memory. The resulting
+    map has one result expression.
+
+    e.g.:
+    ```
+    my_list = [1, 2, 3, 4, 5, 6]
+    shape = [2, 3]
```

Comment:
is the shape `[2, 3]` or `[3, 2]`? Sorry, but the reversing somehow trips me up on what I expect.
Not sure what's actionable here.

Reply:
I agree, keeping it as it is in the other dialects (incl. upstream) would help.

Reply:
The good news is that as long as we get it wrong an even number of times then we'll get the right result in the end :)